File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        seen = set()
        max_len = 0
        start = 0

        for i in range(len(s)):
            while s[i] in seen:
                seen.remove(s[start])
                start +=1
            seen.add(s[i])
            max_len = max(max_len, len(seen))
        return max_len
        

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
import java.util.*;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> seen = new HashMap<>();
        int left = 0;
        int maxLen = 0;
        
        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            // If char seen before, move left pointer past its last index
            if (seen.containsKey(ch) && seen.get(ch) >= left) {
                left = seen.get(ch) + 1;
            }
            seen.put(ch, right);
            maxLen = Math.max(maxLen, right - left + 1);
        }
        
        return maxLen;
    }
}

